plutus-tx/src/PlutusTx/Builtins/HasOpaque.hs

```diff
@@ -0,0 +1,243 @@
+
+{-# OPTIONS_GHC -fno-specialise #-}
+{-# OPTIONS_GHC -fno-omit-interface-pragmas #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module PlutusTx.Builtins.HasOpaque where
+
+import PlutusTx.Base (id, ($))
+import PlutusTx.Bool (Bool (..))
+import PlutusTx.Builtins.Internal
+
+import Data.Kind qualified as GHC
+import Data.String (IsString (..))
+import Data.Text qualified as Text
+import GHC.Magic qualified as Magic
+import Prelude qualified as Haskell (String)
+
+{- Note [noinline hack]
+For some functions we have two conflicting desires:
+- We want to have the unfolding available for the plugin.
+- We don't want the function to *actually* get inlined before the plugin runs, since we rely
+on being able to see the original function for some reason.
+
+'INLINABLE' achieves the first, but may cause the function to be inlined too soon.
+
+We can solve this at specific call sites by using the 'noinline' magic function from
+GHC. This stops GHC from inlining it. As a bonus, it also won't be inlined if
+that function is compiled later into the body of another function.
+
+We do therefore need to handle 'noinline' in the plugin, as it itself does not have
+an unfolding.
+
+Another annoying quirk: even if you have 'noinline'd a function call, if the body is
+a single variable, it will still inline! This is the case for the obvious definition
+of 'stringToBuiltinString' (since the newtype constructor vanishes), so we have to add
+some obfuscation to the body to prevent it inlining.
+-}
+
+obfuscatedId :: a -> a
+obfuscatedId a = a
+{-# NOINLINE obfuscatedId #-}
+
+stringToBuiltinByteString :: Haskell.String -> BuiltinByteString
+stringToBuiltinByteString str = encodeUtf8 $ stringToBuiltinString str
+{-# INLINABLE stringToBuiltinByteString #-}
+
+stringToBuiltinString :: Haskell.String -> BuiltinString
+-- To explain why the obfuscatedId is here
+-- See Note [noinline hack]
+stringToBuiltinString str = obfuscatedId (BuiltinString $ Text.pack str)
+{-# INLINABLE stringToBuiltinString #-}
+
+{- Same noinline hack as with `String` type. -}
+instance IsString BuiltinByteString where
+    -- Try and make sure the dictionary selector goes away, it's simpler to match on
+    -- the application of 'stringToBuiltinByteString'
+    -- See Note [noinline hack]
+    fromString = Magic.noinline stringToBuiltinByteString
+    {-# INLINE fromString #-}
+
+-- We can't put this in `Builtins.hs`, since that force `O0` deliberately, which prevents
+-- the unfoldings from going in. So we just stick it here. Fiddly.
+instance IsString BuiltinString where
+    -- Try and make sure the dictionary selector goes away, it's simpler to match on
+    -- the application of 'stringToBuiltinString'
+    -- See Note [noinline hack]
+    fromString = Magic.noinline stringToBuiltinString
+    {-# INLINE fromString #-}
+
+{- Note [Built-in types and their Haskell versions]
+'HasToBuiltin' allows us to convert a value of a built-in type such as 'ByteString' to its Plutus
```

Comment:
What's the use case for this, I think it is only useful for lifting? Worth elaborating.

Reply:
Lifting is the primary use case for `fromBuiltin`, yes, but we also have `toBuiltin` which is used a lot in benchmarks, for one example (there are many):

```haskell
benchPairing :: EvaluationContext -> Benchmark
benchPairing ctx =
    case listOfSizedByteStrings 4 4 of
      [b1, b2, b3, b4] ->
          let emptyDst = Tx.toBuiltin BS.empty
              p1 = Tx.bls12_381_G1_hashToGroup (Tx.toBuiltin b1) emptyDst
              p2 = Tx.bls12_381_G2_hashToGroup (Tx.toBuiltin b2) emptyDst
              q1 = Tx.bls12_381_G1_hashToGroup (Tx.toBuiltin b3) emptyDst
              q2 = Tx.bls12_381_G2_hashToGroup (Tx.toBuiltin b4) emptyDst
              prog = mkPairingScript p1 p2 q1 q2
          in bench "pairing" $ benchProgramCek ctx prog
      _ -> error "Unexpected list returned by listOfSizedByteStrings"
```

In `PlutusLedgerApi.V1.Bytes`:

```haskell
fromBytes :: BS.ByteString -> LedgerBytes
fromBytes = LedgerBytes . P.toBuiltin

fromHex :: BS.ByteString -> Either LedgerBytesError LedgerBytes
fromHex = fmap (LedgerBytes . P.toBuiltin) . asBSLiteral
```

In `PlutusLedgerApi.V1.Value`:

```haskell
currencySymbol :: BS.ByteString -> CurrencySymbol
currencySymbol = CurrencySymbol . PlutusTx.toBuiltin

tokenName :: BS.ByteString -> TokenName
tokenName = TokenName . PlutusTx.toBuiltin
```

I.e. `toBuiltin` is used for converting Haskell values into into Plutus Tx values.

Unsurprisingly, `fromBuiltin`  is used for converting Plutus Tx values into Haskell values, for example in `PlutusLedgerApi.V1.Bytes`

```haskell
bytes :: LedgerBytes -> BS.ByteString
bytes = P.fromBuiltin . getLedgerBytes
```

In `PlutusLedgerApi.V1.Value`:

```haskell
fromTokenName :: (BS.ByteString -> r) -> (Text -> r) -> TokenName -> r
fromTokenName handleBytestring handleText (TokenName bs) = either (\_ -> handleBytestring $ PlutusTx.fromBuiltin bs) handleText $ E.decodeUtf8' (PlutusTx.fromBuiltin bs)
```

In `PlutusBenchmark.Marlowe.Util`:

```haskell
makeBuiltinData
  :: LedgerBytes
  -> BuiltinData
makeBuiltinData =
  dataToBuiltinData
    . deserialise
    . LBS.fromStrict
    . fromBuiltin
    . getLedgerBytes
```

So I think the current description is as accurate as it gets? `HasFromOpaque`/`HasToOpaque` are trickier and we need to explain what they are for (hence the Note, most of which I took from the existing Michael's docs), but `HasFromBuiltin` and `HasToBuiltin` are just for converting values between Haskell and Plutus Tx.